### PR TITLE
Add support for Helm Upgrade and history

### LINKED ISSUE #446: Helm: Support helm upgrade of a workload

### LINKED ISSUE #460: Helm: Allow users to view revision history for a Helm Release
Allow the user to see the same information as per the Helm CLI, see:  https://helm.sh/docs/helm/helm_history/

### PR DESCRIPTION
**Pull request by nwmac**
fixes #460 
fixes #446 

This PR adds support for upgrading a Helm Release to a new version. It also allows the revision history for a workload to be viewed.

### CODE REVIEW

File: src/frontend/packages/suse-extensions/src/custom/helm/store/helm.action-builders.ts [hunk 1]
```diff
@@ -20,4 +20,12 @@ export interface HelmVersionActionBuilders extends OrchestratedActionBuilders {
 
 export const helmVersionActionBuilders: HelmVersionActionBuilders = {
   getMultiple: () => new GetHelmVersions()
-}
\ No newline at end of file
+}
+
+export interface HelmChartVersionsActionBuilders extends OrchestratedActionBuilders {
+  getMultiple: (repoName: string, chartName: string) => GetHelmChartVersions
```
Comment (richard-cox) [inline src/frontend/packages/suse-extensions/src/custom/helm/store/helm.action-builders.ts:26]: Could possibly use generic `getMultiple`

File: src/frontend/packages/suse-extensions/src/custom/helm/store/helm.types.ts [hunk 2]
```diff
@@ -7,26 +10,18 @@ export interface MonocularRepository {
...
-        version: string
-      }
-    }
-  };
+}
+
+export type MonocularVersion = ChartVersion;
+
+// Basic Chart Metadata
+export interface ChartMetadta {
```
Comment (richard-cox) [inline src/frontend/packages/suse-extensions/src/custom/helm/store/helm.types.ts]: typo

File: src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/release/tabs/helm-release-helper.service.ts [hunk 3]
```diff
@@ -89,6 +134,26 @@ export class HelmReleaseHelperService {
     );
   }
 
+  // Check to see if a workload has updates available
+  public checkForUpdates() {
```
Comment (richard-cox) [inline src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/release/tabs/helm-release-helper.service.ts]: Naming - getCharts/refreshCharts

File: src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/release/tabs/helm-release-helper.service.ts [hunk 4]
```diff
@@ -89,6 +134,26 @@ export class HelmReleaseHelperService {
...
+    helmEntityCatalog.chart.api.getMultiple();
+    return helmEntityCatalog.chart.store.getPaginationMonitor().currentPage$.pipe(
```
Comment (richard-cox) [inline src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/release/tabs/helm-release-helper.service.ts]: Remove api.getMultiple, change line 140 to getPaginationService

File: src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/release/tabs/helm-release-helper.service.ts [hunk 5]
```diff
@@ -89,6 +134,26 @@ export class HelmReleaseHelperService {
...
+      filter(charts => !!charts)
+    );
+  }
+
+  public fetchReleaseHistory(): Observable<any> {
+    // Get the history for a Helm release
+    workloadsEntityCatalog.history.api.get(this.releaseTitle, this.endpointGuid, { namespace: this.namespace });
+    const action = workloadsEntityCatalog.history.actions.get(this.releaseTitle, this.endpointGuid, { namespace: this.namespace });
+    return workloadsEntityCatalog.history.store.getEntityMonitor(
```
Comment (richard-cox) [inline src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/release/tabs/helm-release-helper.service.ts]: workloadsEntityCatalog.history.store.getEntityService(..).waitForEntity$

File: src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/release/tabs/helm-release-helper.service.ts [hunk 6]
```diff
@@ -132,15 +197,62 @@ export class HelmReleaseHelperService {
     };
   }
 
-  private isContainerReady(state: ContainerStateCollection = {}): Boolean {
+  private isContainerReady(state: ContainerStateCollection = {}): boolean {
```
Comment (richard-cox) [inline src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/release/tabs/helm-release-helper.service.ts]: Linting needs disabling, should remain as Boolean

File: src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/release/tabs/helm-release-helper.service.ts [hunk 7]
```diff
@@ -132,15 +197,62 @@ export class HelmReleaseHelperService {
...
       return false;
     } else if (!!state.terminated) {
       // Assume a failed state is not ready (covers completed init states), discard success state
-      return state.terminated.exitCode === 0 ? null : false
+      return state.terminated.exitCode === 0 ? null : false;
     }
     return false;
   }
+
+  public hasUpgrade(): Observable<any> {
```
Comment (richard-cox) [inline src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/release/tabs/helm-release-helper.service.ts]: typing

File: src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/release/tabs/helm-release-helper.service.ts [hunk 8]
```diff
@@ -132,15 +197,62 @@ export class HelmReleaseHelperService {
...
+        return null;
+      })
+    );
+  }
+
+  // We might have a chart with the same name in multiple repositories - we only have chart metadata
+  // We don't know which Helm repository it came from, so use the description and sources to match
+  private isProbablySameChart(a: ChartMetadta, b: ChartMetadta): boolean {
+    // Basic properties must be the same
+    if ((a.name !== b.name) ||  (a.description !== b.description) || (a.sources.length !== b.sources.length)) {
```
Comment (richard-cox) [inline src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/release/tabs/helm-release-helper.service.ts]: Remove `description`

File: src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/release/tabs/helm-release-helper.service.ts [hunk 9]
```diff
@@ -139,8 +202,55 @@ export class HelmReleaseHelperService {
...
+        return null;
+      })
+    );
+  }
+
+  // We might have a chart with the same name in multiple repositories - we only have chart metadata
+  // We don't know which Helm repository it came from, so use the description and sources to match
+  private isProbablySameChart(a: ChartMetadta, b: ChartMetadta): boolean {
+    // Basic properties must be the same
+    if ((a.name !== b.name) || (a.description !== b.description) || (a.sources.length !== b.sources.length)) {
```
Comment (richard-cox) [inline src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/release/tabs/helm-release-helper.service.ts]: description can be removed

File: src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/upgrade-release/upgrade-release.component.html [hunk 10]
```diff
@@ -0,0 +1,37 @@
+<app-page-header>
+  Upgrade Workload
+</app-page-header>
+<app-steppers [cancel]="cancelUrl">
+  <app-step title="Version" [valid]="validate$ | async">
+    <app-list class="workload-upgrade__list" *ngIf="listConfig" [listConfig]="listConfig"></app-list>
```
Comment (richard-cox) [inline src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/upgrade-release/upgrade-release.component.html:6]: If there's no upgrades this is just a blank step.

File: src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/release/tabs/helm-release-helper.service.ts [hunk 11]
```diff
@@ -71,7 +116,7 @@ export class HelmReleaseHelperService {
 
   public fetchReleaseResources(): Observable<HelmReleaseResources> {
     // Get helm release
-    const action = workloadsEntityCatalog.resource.actions.get(this.releaseTitle, this.endpointGuid)
+    const action = workloadsEntityCatalog.resource.actions.get(this.releaseTitle, this.endpointGuid);
```
Comment (richard-cox) [inline src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/release/tabs/helm-release-helper.service.ts]: action can be deleted

File: src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/store/workloads-entity-generator.ts [hunk 12]
```diff
@@ -7,22 +7,35 @@ import { IFavoriteMetadata } from '../../../../../../store/src/types/user-favori
...
+  helmReleaseGraphEntityType,
+  helmReleaseHistoryEntityType,
+  helmReleaseResourceEntityType,
+  registerWorkloadEntityCatalog,
+} from './workloads-entity-factory';
 
 
 export function generateWorkloadsEntities(endpointDefinition: StratosEndpointExtensionDefinition): StratosBaseCatalogEntity[] {
+
+  registerWorkloadEntityCatalog();
```
Comment (richard-cox) [inline src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/store/workloads-entity-generator.ts]: double check

File: src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/store/workloads.effects.ts [hunk 13]
```diff
@@ -75,14 +86,84 @@ export class WorkloadsEffects {
           } as NormalizedResponse;
 
           // Go through each endpoint ID
-          processedData.entities[entityKey][action.guid] = this.mapHelmRelease(response, action.endpointGuid, action.guid);
+          processedData.entities[entityKey][action.guid] = response;
```
Comment (richard-cox) [inline src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/store/workloads.effects.ts]: Double check removal of mapHelmRelease

File: src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/workload.types.ts [hunk 14]
```diff
@@ -67,6 +71,19 @@ export interface HelmReleaseResources extends HelmReleaseEntity {
   kind: string
 };
 
+export interface HelmReleasRevision {
```
Comment (richard-cox) [inline src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/workload.types.ts]: typo

File: src/jetstream/plugins/kubernetes/install_release.go [hunk 15]
```diff
@@ -170,11 +181,79 @@ func (c *KubernetesSpecification) DeleteRelease(ec echo.Context) error {
...
+	buf := new(bytes.Buffer)
+	buf.ReadFrom(bodyReader)
+
+	var params upgradeRequest
+	err := json.Unmarshal(buf.Bytes(), &params)
+	if err != nil {
+		return interfaces.NewJetstreamErrorf("Could not get Upgrade Release Parameters: %+v", err)
+	}
+
+	config, hc, err := c.GetHelmConfiguration(endpointGUID, userGUID, namespace)
```
Comment (richard-cox) [inline src/jetstream/plugins/kubernetes/install_release.go:234]: RC Test with helm hub

**Review by richard-cox — CHANGES_REQUESTED** [review]
See comments

File: src/frontend/packages/suse-extensions/src/custom/helm/store/helm.types.ts (hunk 2 above)
Comment (nwmac) [reply src/frontend/packages/suse-extensions/src/custom/helm/store/helm.types.ts]: Fixed

File: src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/store/workloads-entity-generator.ts (hunk 12 above)
Comment (nwmac) [reply src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/store/workloads-entity-generator.ts]: This doesn't seen to be needed, so I have reverted it.

File: src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/store/workloads.effects.ts (hunk 13 above)
Comment (nwmac) [reply src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/store/workloads.effects.ts]: Checked this and reverted back

File: src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/workload.types.ts (hunk 14 above)
Comment (nwmac) [reply src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/workload.types.ts]: Fixed

File: src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/release/tabs/helm-release-helper.service.ts [hunk 16]
```diff
@@ -227,14 +225,23 @@ export class HelmReleaseHelperService {
             }
           }
         }
+        // No newer release, so return the release itself if that is what was requested
+        if (returnLatest) {
+          const releaseChart = charts.find(c => c.relationships.latestChartVersion.data.version === release.chart.metadata.version);
+          return {
+            release,
+            upgrade: releaseChart.attributes,
```
Comment (richard-cox) [inline src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/release/tabs/helm-release-helper.service.ts]: If the workload was not deployed via a connected helm, or that helm was removed, this results in a `Cannot read property of undefined ` error. Feels like the returnLatest logic should be moved into the `if (latest.isNewer(current)) {` check and the case where there's no current or newer chart handled here

Comment (nwmac) [reply src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/release/tabs/helm-release-helper.service.ts]: Fixed and updated so the upgrade button only shows if there is an available helm chart

**Review by richard-cox — APPROVED** [review]

**Pull request merged**